docs/data/data-grid/pagination/pagination.md

```diff
@@ -94,40 +94,155 @@ const [paginationModel, setPaginationModel] = React.useState({
+    rowCountRef.current = pageInfo.totalRowCount;
+  }
+  return rowCountRef.current;
+}, [pageInfo?.totalRowCount]);
+
+<DataGrid rowCount={rowCount} />;
+```
+
 {{"demo": "ServerPaginationGrid.js", "bg": "inline"}}
 
-### Cursor implementation
+#### Unknown row count
+
+Pass the props to the Grid as explained in the table above to handle the case when the actual row count is not initially available.
+
+The value of the `hasNextPage` variable might become `undefined` during loading if it's handled by some external fetching hook resulting in unwanted computations, one possible solution could be to memoize the `paginationMeta`:
+
+```tsx
+const paginationMetaRef = React.useRef<GridPaginationMeta>();
+
+const paginationMeta = React.useMemo(() => {
+  if (
+    hasNextPage !== undefined &&
+    paginationMetaRef.current?.hasNextPage !== hasNextPage
+  ) {
+    paginationMetaRef.current = { hasNextPage };
+  }
+  return paginationMetaRef.current;
+}, [hasNextPage]);
+```
+
+{{"demo": "ServerPaginationGridNoRowCount.js", "bg": "inline"}}
+
+#### Estimated row count
+
+There could be possibilities when the accurate row count is not initially available for many reasons such as:
+
+1. For some databases, computing `rowCount` upfront is a costly operation due to the scale of data or how the data is structured.
+2. Some data structures don't have the `rowCount` information until the very last page.
+
+In such scenarios, some backends provide an estimated row count initially which could be used to estimate the number of pages until the actual row count is available. In some cases, this estimate value could also be more than the actual row count. Therefore, the Grid uses it only to show the user an estimated number of rows, and uses `paginationMeta.hasNextPage` as a single source of truth to check whether more records are available on the server.
+
+There could be two possibilities after providing the initial props (as mentioned in the table above):
+
+1. The actual row count is fetched from the server lazily and provided to the Grid using the `rowCount` prop. The `estimatedRowCount` prop will be ignored once the `rowCount` prop is set to a positive value.
+2. The user has already reached the estimated last page and the actual row count is still not available, in that case, the Grid could take the help of the `hasNextPage='false'` to know that the last page is fetched and try to set a value like in the unknown row count use case.
+
+The following example demonstrates the use of the `estimatedRowCount` and `hasNextPage` props to handle the case when the actual row count is not initially available. The actual row count is `1000` but the Grid is initially provided with an estimated row count of `100`.
+
+The Grid keeps fetching the next page until `hasNextPage` is `false` or the actual row count is provided to the Grid lazily, you can do that by clicking the "Set Row Count" button.
+
+{{"demo": "ServerPaginationGridEstimated.js", "bg": "inline"}}
```

Comment:
This might be out of this PR's scope, but did you consider custom pagination components in this scenario?
E.g. the one with individual page buttons in https://mui.com/x/react-data-grid/components/#pagination

In this particular example, it looks wrong: https://codesandbox.io/p/sandbox/inspiring-cherry-svvzh6

<img width="842" src="https://github.com/mui/mui-x/assets/13808724/ed65efba-8668-4681-a1c7-05fe579cb388">

- It shows only page 1 (although `estimatedRowCount` could be utilized here)
- And more importantly, the arrow buttons are disabled. This seems to be caused by https://github.com/mui/mui-x/pull/12474/files?w=1#diff-79d72ca24d1e0278032ce56430db49ef25dc59fd363d0d063912654cddc850ceR14

Reply:
Here's the updated demo: https://codesandbox.io/p/sandbox/affectionate-violet-qfjdgw?file=%2Fsrc%2FDemo.tsx